jetty-openid/src/main/java/org/eclipse/jetty/security/openid/OpenIdConfiguration.java

```diff
@@ -69,60 +71,94 @@ public OpenIdConfiguration(String provider, String clientId, String clientSecret
      * @param tokenEndpoint the URL of the OpenID provider's token endpoint if configured.
      * @param clientId OAuth 2.0 Client Identifier valid at the Authorization Server.
      * @param clientSecret The client secret known only by the Client and the Authorization Server.
+     * @param httpClient The {@link HttpClient} instance to use.
      */
-    public OpenIdConfiguration(String issuer, String authorizationEndpoint, String tokenEndpoint, String clientId, String clientSecret)
+    public OpenIdConfiguration(String issuer, String authorizationEndpoint, String tokenEndpoint,
+                               String clientId, String clientSecret, HttpClient httpClient)
     {
         this.issuer = issuer;
         this.clientId = clientId;
         this.clientSecret = clientSecret;
+        this.authEndpoint = authorizationEndpoint;
+        this.tokenEndpoint = tokenEndpoint;
+        this.httpClient = httpClient;
 
-        if (issuer == null)
-            throw new IllegalArgumentException("Provider was not configured");
+        if (this.issuer == null)
+            throw new IllegalArgumentException("Issuer was not configured");
 
-        if (tokenEndpoint == null || authorizationEndpoint == null)
+        addBean(this.httpClient);
+    }
+
+    @Override
+    protected void doStart() throws Exception
+    {
+        super.doStart();
+
+        if (authEndpoint == null || tokenEndpoint == null)
         {
-            Map<String, Object> discoveryDocument = fetchOpenIdConnectMetadata(issuer);
+            Map<String, Object> discoveryDocument = fetchOpenIdConnectMetadata(issuer, httpClient);
 
-            this.authEndpoint = (String)discoveryDocument.get("authorization_endpoint");
-            if (this.authEndpoint == null)
+            authEndpoint = (String)discoveryDocument.get("authorization_endpoint");
+            if (authEndpoint == null)
                 throw new IllegalArgumentException("authorization_endpoint");
 
-            this.tokenEndpoint = (String)discoveryDocument.get("token_endpoint");
-            if (this.tokenEndpoint == null)
+            tokenEndpoint = (String)discoveryDocument.get("token_endpoint");
+            if (tokenEndpoint == null)
                 throw new IllegalArgumentException("token_endpoint");
 
             if (!Objects.equals(discoveryDocument.get("issuer"), issuer))
-                LOG.warn("The provider in the metadata is not correct.");
-        }
-        else
-        {
-            this.authEndpoint = authorizationEndpoint;
-            this.tokenEndpoint = tokenEndpoint;
+                throw new IllegalArgumentException("The provider in the metadata is not correct.");
```

Comment:
I think a throw is too dramatic. If the OP is misconfigured, we have all the other config that we need (the endpoints). If we just warn, we may be able to use the OP even though it's oddly configured. With this throw, we're broken as well.

Reply:
For the validation of the ID Token it says `The Issuer Identifier for the OpenID Provider (which is typically obtained during Discovery) MUST exactly match the value of the iss (issuer) Claim.`

But if we have two values of this then which one do we use to get an exact match. This seems like bad configuration to me.

Reply:
We match config and ignore the metadata results. I know we're bending the rules here a bit, but I think the OIDC spec is being too strict if it requires us to throw. We have what we need and might be able to succeed to interoperate. The reason I'm suggesting to be more tolerant is because of this:

> be conservative in what you do, be liberal in what you accept from  others.

https://tools.ietf.org/html/rfc793#section-2.10

There's a standard Relying Party conformance test about this that accepts logging as good enough IIRC. My idea is to run those and submit Jetty once we get a little further along. When we do, we might revisit this then.

Reply:
Not convinced that this couldn't be a security problem that someone who doesn't identify as the issuer we expected is telling us the location of the tokenEndpoint and authEndpoint. At best it is probably likely to just fail when authenticating when they send ID Tokens with the same issuer they specified in the metadata.

We always have the option to enter in the authEndpoint and tokenEndpoint manually, so maybe it is best to leave this and see if we encounter them.

Reply:
We authenticated the server with SSL as some entity that we trusted. So, we're either being man-in-the-middled or the OP's metadata has been poisoned. Later, when the OP issues an ID token, we'll reject it if its issuer doesn't match the config one. If we're really scared of this, we should be verifying the signature of the ID token as well. I think that this is not a risk and something that we should do in the name of interop.

Reply:
I have even found an example of this in Microsoft's OIDC service so maybe it would be better just to warn.
https://login.microsoftonline.com/common/v2.0/.well-known/openid-configuration

I have changed it back to a warning.

Reply:
We also allow the issuer to be overridden to be whatever is needed in a deployment. It doesn't come up often, but seems useful in migrations. That multi-tenancy use case is new to me; thanks for sharing that.